packages/core/nano/src/types.ts

```diff
@@ -1,22 +1,139 @@
+ * a reference, through getDataResult. Formula is expected to implement IConsumer
+ * and register itself as an observer to Tablero.
+ *
+ * Any component that implements IProducer is expected to provide some registration
+ * functionality and to notify observers whenever the data they are bound to changes.
+ */
+export interface IProducer<T> {
+  /**
+   * Binds the given observer to the component associated with this Producer.
+   * @param consumer - The object to start listening to any changes in the bound data.
+   */
+  addConsumer(consumer: IConsumer<T>): void;
+
+  /**
+   * Unbinds the given observer from the component associated with this Producer.
+   * @param consumer - The observer to unregister from the Producer.
+   */
+  removeConsumer(consumer: IConsumer<T>): void;
+
+  /**
+   * Return IDataResult for the given dataBindingReference, containing the data value
+   * and some additional metadata.
+   * @param dataBindingReference - An instance of dataBindingReference, a reference
+   * to some data in the bound component.
+   */
+  read<K extends keyof T>(consumer: IConsumer<T>, property: K, ...args: any[]): T[K] | Pending<T[K]>;
+
+  /**
+   * Invoked when this data provider is deleted.
+   */
+  onDeleted(): void;
+}
+
+/**
+ * The interface for an object that can bind to an IProducer.
+ * One example of this provider-observer relationship is a Formula binding
+ * to Tablero. In this scenario, Formula is an IConsumer listening to changes
+ * in the IProducer, Tablero.
+ *
+ * Any object that implements IConsumer is expected to provide a
+ * callback whenever the component it is bound to changes in value and a reference
+ * to the data that the observer is bound to.
+ */
+export interface IConsumer<T> {
+  /**
+   * Invoked whenever the data this object is bound to is changed.
+   */
+  notify: <K extends keyof T>(property: K, value: T[K]) => void;
+
+  /**
+   * Invoked whenever the provider this is bound to is deleted.
+   */
+  onProducerDeleted: () => void;
 }
 
-export type PickPending<T, K extends keyof T> = { [Key in K]: T[K] | Pending<T[K]> };
-
-export interface Producer<T> {
-    id: string;
-    unsubscribe: (origin: Consumer<T>) => void;
-    
-    now: <K extends keyof T>(property: K, ...args: any[]) => T[K] | undefined;
-    nowMany: <K extends keyof T>(properties: Record<K, undefined | any[]>) => Partial<Pick<T, K>>;
-    
-    request: <K extends keyof T>(origin: Consumer<T>, property: K, ...args: any[]) => T[K] | Pending<T[K]>;
-    requestMany: <K extends keyof T>(origin: Consumer<T>, properties: Record<K, undefined | any[]>) => PickPending<T,K>;
+/**
+ * IDataResult represents the result object passed through data binding between two components.
+ * It provides query functions for retrieving values, hints, friendly names, and provides
+ * basic traverse APIs to move to previous or next level results.
+ */
+export interface IDataResult {
```

Comment:
I'm not yet sure how to reconcile IDataResult.

I'm currently thinking that the Intellisense related APIs should be a separate interface at the framework level that an IProducer optionally implements, like:

```ts
class Producer implements IProducer<T>, IIntellisenseProvider<T> {
    getProperties(query: string) => Promise<string[]>;
    getHint(property: string): string;
    getFriendlyName(property: string): string;

    // ..plus regular IProducer stuff
}
```
...unless Jack has any use for these APIs?

Reply:
If my understanding is correct `IDataResult` is a static value and does not contain any observables/producers within it - in that sense `IDataResult` is really just an app specific value that a `IProducer` may provide. For the `IProducer` type I think there are two constraints we want to juggle (there may be more). First, a conceptually simple producer should be simple to implement, so we don't want to make the base interfaces too heavy. Second, I think there should be some standard way to explore a network of producers.

I'm inclined to say that `IProducer` should have the method `getProperties(query: string) => Promise<string[]>;` with the contract that the array of strings is a lower bound on the set of subscribable properties (perhaps even make the method optional). This way someone could quickly implement something like `getProperties() => Promise.resolve([])` without violating the contract.

I think the other things would belong in supplementary interfaces as @DLehenbauer suggests.

Reply:
And I don't have any uses for these API's atm. Those might likely one I would need is a way to enumerate properties and possibly get their types, but that is further down the line.